### DISCUSSION TITLE
Testing type predicates

### DISCUSSION THREAD

**Discussion by olsonpm** [tstyche/tstyche#303] [Q&A] [answered]
Thanks for this tool, it's helping me a lot.

I ran into unexpected behavior when explicitly throwing an error in an 'else' statement at the end of a test - the test still succeeded.

*I understand I may be using this tool in an unintended manner, however I believe the behavior is still unintuitive*

Here's a minimal reproduction

```sh
$ git clone git@github.com:olsonpm/repro.git
$ cd repro
$ git checkout tstyche-error-passes
$ npm ci
$ npx tstyche
# test will pass when I would expect it to fail based off the last line in `MethodLikeKeys.tst.ts`
```

Is this behavior by design ?  I'm willing to create a PR if not

---

edit: for anyone encountering this, my workaround is `expect('failure explanation here').type.toBeAny()`

**Comment by mrazauskas** [comment, marked as answer]
I took a quick look. All works as expected.

TSTyche does not execute the code. It only performs static analysis which is identical to the analysis performed by `tsc`.

Since typings are correct and no syntax errors are found in the file, it is passing. Or did I miss something?

Could you tell more what you are trying to test? I can help to figure out a better solution.

**Reply by olsonpm** [reply]
> TSTyche does not execute the code

Gotcha, and interesting - so the function calls are all for familiarity ?  e.g. describe(), test(), expect() ?  I'll dig around your code to better understand.

> Could you tell more about what you are trying to test ?

sure, I'm trying to test [type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) behavior for a function `negate`.  

e.g.

```ts
import negate from 'path/to/negate'

type Dog = { name: 'dog' }
type Fox = { name: 'fox' }
type Animal = Dog | Fox

const isDog = (animal: Animal): animal is Dog => animal.name === 'dog'
const isFox = negate(isDog)
const animal: Animal = { name: 'fox' }

// type predicate testing here
if (isFox(animal)) expect(animal).type.toBe<Fox>()
else throw new Error('animal must be Fox here')
```

I appreciate your time

**Reply by olsonpm** [reply]
I just realized I should simplify the test to be

```ts
expect(isFox).type.toBe<(data: Animal) => data is Fox>()
```

not sure why my first instinct was to test the predicate scoping itself.

sorry for the bother

**Reply by mrazauskas** [reply]
I think it was good idea to use `if` in this case. Only the `else` branch is redundant. It can be removed.

For example, a test for `isFish()` from the TypeScript docs, could look like this:

```ts
if (isFish(pet)) {
  expect(pet).type.toBe<Fish>();
} else {
  expect(pet).type.toBe<Bird>();
}
```

Here behaviour of `isFish()` is tested with different input. This is similar to functional testing: given the input, what is the output?

**Reply by mrazauskas** [reply]
You are right, `describe()`, `test()` and `expect()` are designed to look familiar. They also have `.only`, `.skip`, `.todo` modifiers. These are priceless when you have to debug something (`.skip` also skips any type errors, of course).

Names of test cases are printed out in the output. And the callback function isolates context.

One more detail: `expect()` also has `.only` and `.skip` modifiers.

**Reply by olsonpm** [reply]
gotcha, thanks much for explaining
